cmd/acraserver/listener.go

```diff
@@ -107,30 +145,149 @@ func (server *SServer) handleConnection(connection net.Conn) {
 
 // start listening connections from proxy
 func (server *SServer) Start() {
-	listener, err := network.Listen(server.config.GetAcraConnectionString())
+	var connection net.Conn
+	var listener, err = network.Listen(server.config.GetAcraConnectionString())
 	if err != nil {
 		log.WithError(err).Errorln("can't start listen connections")
+		ErrorSignalChannel <- syscall.SIGTERM
 		return
 	}
-	defer listener.Close()
+	server.listenerACRA = listener
 	server.addListener(listener)
+
+	log.Infof("start listening %s", server.config.GetAcraConnectionString())
+	for {
+		connection, err = listener.Accept()
+		if err != nil {
+			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+				log.Infoln("stop accepting new connections due net.Timeout")
+				return
+			}
+			log.WithError(err).Errorf("can't accept new connection (connection=%v)", connection)
+		}
+		// unix socket and value == '@'
+		if len(connection.RemoteAddr().String()) == 1 {
+			log.Infof("new connection to acraserver: <%v>", connection.LocalAddr())
+		} else {
+			log.Infof("new connection to acraserver: <%v>", connection.RemoteAddr())
+		}
+		go func() {
+			server.cmACRA.Add(1)
+			server.handleConnection(connection)
+			server.cmACRA.Done()
+		}()
+
+	}
+}
+
+func (server *SServer) StartFromFileDescriptor(fd uintptr) {
+	var connection net.Conn
+	file := os.NewFile(fd, "/tmp/acraserver")
+	listenerFile, err := net.FileListener(file)
+	if err != nil {
+		log.WithError(err).Errorln("can't start listen for file descriptor")
+		ErrorSignalChannel <- syscall.SIGTERM
+		return
+	}
+
+	listenerWithFileDesciptor, ok := listenerFile.(network.ListenerWithFileDescriptor)
+	if !ok {
+		log.WithError(err).Errorf("File descriptor %d is not a valid socket", fd)
+		return
+	}
+	server.listenerACRA = listenerWithFileDesciptor
+	server.addListener(listenerFile)
+
 	log.Infof("start listening %s", server.config.GetAcraConnectionString())
 	for {
-		connection, err := listener.Accept()
+		connection, err = listenerWithFileDesciptor.Accept()
 		if err != nil {
-			log.WithError(err).Errorln("can't accept new connection")
-			continue
+			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+				log.WithError(err).Errorf("stop accepting new connections", connection)
+				return
+			}
+			log.WithError(err).Errorf("can't accept new connection (connection=%v)", connection)
 		}
 		// unix socket and value == '@'
 		if len(connection.RemoteAddr().String()) == 1 {
 			log.Infof("new connection to acraserver: <%v>", connection.LocalAddr())
 		} else {
 			log.Infof("new connection to acraserver: <%v>", connection.RemoteAddr())
 		}
-		go server.handleConnection(connection)
+		go func() {
+			server.cmACRA.Add(1)
+			server.handleConnection(connection)
+			server.cmACRA.Done()
+		}()
+	}
+}
+
+func (server *SServer) StopListeners() {
+	var (
+		err  error
+		nerr *net.TCPListener
```

Comment:
maybe better to name variable as `tcpListener`? when I read first, I thought that it `network error`... and wanted to say that you don't need to declare such variable here because you use it in `if block`. But then found that it not error)

Reply:
done